File: spa/src/ui/units.test.ts
import { describe, expect, it } from 'vitest';
import { CONFORMANCE, LABEL_CONFORMANCE } from './units.conformance.ts';
import {
  CANONICAL_LOCALE,
  DEGREES,
  exactValue,
  formatNumber,
  formatValue,
  GS,
  JOULES,
  KILOGRAMS,
  METRES,
  METRES_SEC,
  METRES_SEC2,
  PASCALS,
  SECONDS,
  unitForKey,
  unitLabel,
  unitMeasured,
} from './units.ts';

/**
 * The whole point of `units.ts`.
 *
 * `units.Conformance` is the contract the Go implementation in
 * `server/internal/units` is held to by its own `units_test.go`; this asserts
 * the TypeScript half of the same table. A failure here means one of two things
 * happened: either a rule changed — in which case the Go package comment, its
 * table and this copy all move together — or a rule broke, in which case
 * catlog's two frontends have just started disagreeing about what a record says.
 */
describe('the shared conformance table', () => {
  it.each(CONFORMANCE.map((row) => [row.value, row.unit, row.want] as const))(
    'Format(%p, %p) = %p',
    (value, unit, want) => {
      // Pinned to the canonical locale: the table is the cross-language
      // contract, and a table that read the machine's locale would pass or fail
      // depending on whose machine ran it.
      expect(formatValue(value, unit, CANONICAL_LOCALE)).toBe(want);
    },
  );

  it('reproduces every row the Go table carries', () => {
    // A guard on the transcription itself: rows get deleted by accident during a
    // merge far more easily than they get changed.
    expect(CONFORMANCE).toHaveLength(43);
  });
});

describe('the shared label table', () => {
  it.each(LABEL_CONFORMANCE.map((row) => [row.unit, row.label, row.measured] as const))(
    'Label(%p) = %p, Measured = %p',
    (unit, label, measured) => {
      expect(unitLabel(unit)).toBe(label);
      expect(unitMeasured(unit)).toBe(measured);
    },
  );

  it('reproduces every row the Go table carries', () => {
    expect(LABEL_CONFORMANCE).toHaveLength(16);
  });

  it('names a unit only when every cell in the column ends in it', () => {
    // The whole rule, checked rather than asserted: render a value in the unit
    // and look at what the string ends with. An SI prefix goes *before* the
    // symbol — "1.82 Mm" still ends in "m" — which is why a scaled column keeps
    // its base symbol and a duration column cannot.
    for (const row of LABEL_CONFORMANCE) {
      if (row.unit === '') continue;
      for (const value of [0.45, 37.5, 313, 3661, 90_000, 1234.5, 1.82e6]) {
        const cell = formatValue(value, row.unit);
        expect(cell.endsWith(row.label)).toBe(row.label !== 'Time');
      }
    }
  });

  it('is the fix for a duration column headed "ms"', () => {
    // The defect, spelled out. A career-time board publishes `unit: "ms"` and
    // renders these three strings; not one of them says "ms".
    expect(formatValue(37_500, 'ms')).toBe('37.5 s');
    expect(formatValue(37_380_000, 'ms')).toBe('10h 23m');
    expect(formatValue(2.1e10, 'ms')).toBe('243d 01h');
    expect(unitLabel('ms')).toBe('Time');
    // And the storage unit is not lost — it moves into the prose above the board.
    expect(unitMeasured('ms')).toBe('ms, shown as a duration');
  });
});

describe('grouping', () => {
  it("is the reader's, not a separator catlog picked", () => {
    // The defect this replaced: every reader on earth got a U+202F narrow
    // no-break space between the groups. The joke is that U+202F is not a bad
    // separator — it is fr-FR's, which is the last assertion here — it was just
    // being shown to the other several hundred locales as well.
    expect(formatValue(1234567, '', CANONICAL_LOCALE)).toBe('1,234,567');
    expect(formatValue(1234567, '', 'de-DE')).toBe('1.234.567');
    expect(formatValue(1234567, '', 'en-GB')).toBe('1,234,567');
    expect(formatValue(1234567, '', 'fr-FR')).toBe('1\u202F234\u202F567');
  });

  it('is grouping, not a character swap', () => {
    // Why this cannot be "replace one separator with another" on the rendered
    // string: two locales below disagree about where the groups *are*, not
    // about what goes between them.
    expect(formatNumber(1234567, 'en-IN')).toBe('12,34,567');
    // es-ES leaves a bare four-digit number ungrouped.
    expect(formatNumber(1234, 'es-ES')).toBe('1234');
    expect(formatNumber(1234, CANONICAL_LOCALE)).toBe('1,234');
  });

  it('localises the decimal separator with it', () => {
    expect(formatValue(1820000, 'm', CANONICAL_LOCALE)).toBe('1.82 Mm');
    expect(formatValue(1820000, 'm', 'de-DE')).toBe('1,82 Mm');
    // Rule 5's sub-minute forms are still one number, so they localise too.
    expect(formatValue(37500, 'ms', 'de-DE')).toBe('37,5 s');
    // Its two-component forms are not, and must not acquire a separator.
    expect(formatValue(2.1e7, 's', 'de-DE')).toBe('243d 01h');
  });

  it('keeps the trailing-zero trim that rule 2 asks for', () => {
    // `minimumFractionDigits: 0`, checked: Intl's default of 3 maximum / 0
    // minimum is not what this needs, and a minimum of two would render 1.8
    // as "1.80".
    expect(formatNumber(1.8, CANONICAL_LOCALE)).toBe('1.8');
    expect(formatNumber(1.8, 'de-DE')).toBe('1,8');
    expect(formatNumber(0.002, CANONICAL_LOCALE)).toBe('0.002');
  });

  it('never renders a negative zero', () => {
    // A value that rounds away to nothing keeps no sign: negative zero is a
    // rounding artefact, never a fact about a flight.
    expect(formatNumber(-0, CANONICAL_LOCALE)).toBe('0');
    expect(formatNumber(-1e-9, CANONICAL_LOCALE)).toBe('0');
    expect(formatNumber(-1e-9, 'de-DE')).toBe('0');
    // A value that survives the rounding keeps it.
    expect(formatNumber(-0.004, CANONICAL_LOCALE)).toBe('-0.004');
  });
});

describe('unitForKey', () => {
  it.each([
    // The trap: `_ms` is metres per second in a payload, but the board unit
    // string "ms" is milliseconds. Only this function knows the difference.
    ['speed_ms', METRES_SEC],
    ['surface_speed_ms', METRES_SEC],
    ['fastest_ms', METRES_SEC],
    ['accel_ms2', METRES_SEC2],
    ['altitude_m', METRES],
    ['ap_m', METRES],
    ['travelled_m', METRES],
    ['energy_j', JOULES],
    ['peak_q_pa', PASCALS],
    ['max_q_pa', PASCALS],
    ['dyn_pressure_pa', PASCALS],
    ['duration_s', SECONDS],
    ['mission_time_s', SECONDS],
    ['t1_sim', SECONDS],
    ['sim_t', SECONDS],
    ['mass_kg', KILOGRAMS],
    ['inc_deg', DEGREES],
    ['peak_g', GS],
    // Unrecognised keys get no unit rather than a wrong one.
    ['body', ''],
    ['career', ''],
    ['ecc', ''],
    ['n', ''],
  ])('maps %p to %p', (key, want) => {
    expect(unitForKey(key)).toBe(want);
  });

  it('renders a roster speed as a speed, not as half a minute', () => {
    // The bug this table exists to prevent, spelled out: 30 km/s is an ecliptic
    // -frame roster speed, and a renderer that read `_ms` as milliseconds would
    // print "30 s".
    expect(formatValue(30000, unitForKey('fastest_ms'), CANONICAL_LOCALE)).toBe('30,000 m/s');
    expect(formatValue(30000, 'ms', CANONICAL_LOCALE)).toBe('30 s');
  });
});

describe('every unit a board can publish', () => {
  it.each([
    'm/s',
    'g',
    'tumbles',
    'RUDs',
    'orbits',
    'bodies',
    'dockings',
    'stagings',
    'kittens',
    'm',
    'ms',
  ])('renders %p without falling through to something that looks like a bug', (unit) => {
    const got = formatValue(1234.5, unit);
    expect(got).not.toBe('');
    expect(got.endsWith(' ')).toBe(false);
  });

  it('renders a label this build has never heard of', () => {
    // A board added later gets "12 whatevers", not a blank cell.
    expect(formatValue(12, 'whatevers')).toBe('12 whatevers');
  });
});

describe('exactValue', () => {
  it('keeps the digits the rendered form hides', () => {
    expect(formatValue(48_000_000, 'J')).toBe('48 MJ');
    expect(exactValue(48_000_000, 'J')).toBe('48000000 J');
    expect(exactValue(7799, '')).toBe('7799');
  });

  it('says so rather than putting NaN in a tooltip', () => {
    expect(exactValue(Number.NaN, 'm')).toBe('not a number');
  });
});

File: server/internal/units/units_test.go
package units_test

import (
	"math"
	"strconv"
	"strings"
	"testing"

	"github.com/meow-sci/catlog/server/internal/units"
)

// TestConformance is the whole point of the package: [units.Conformance] is the
// written contract for what a catlog number looks like, and this asserts that
// the code still meets it.
//
// A failure here means one of two things happened. Either a rule changed — in
// which case the package comment and this table move together — or a rule
// broke, in which case a record has just started reading differently than the
// contract says it should.
func TestConformance(t *testing.T) {
	for _, tc := range units.Conformance {
		got := units.Format(tc.Value, tc.Unit)
		if got != tc.Want {
			t.Errorf("Format(%v, %q) = %q, want %q", tc.Value, tc.Unit, got, tc.Want)
		}
	}
}

// TestLabelConformance is the same contract for rule 7. A column of durations
// under a header reading "ms" was the defect this table exists to stop coming
// back: no cell in that column says "ms", so the header names the quantity.
func TestLabelConformance(t *testing.T) {
	for _, tc := range units.LabelConformance {
		if got := units.Label(tc.Unit); got != tc.Label {
			t.Errorf("Label(%q) = %q, want %q", tc.Unit, got, tc.Label)
		}
		if got := units.Measured(tc.Unit); got != tc.Measured {
			t.Errorf("Measured(%q) = %q, want %q", tc.Unit, got, tc.Measured)
		}
	}
}

// A header may only name a unit that a reader will actually see at the end of
// every cell beneath it. That is the whole rule, and it is checkable: render a
// value in the unit and look at what the string ends with.
func TestLabelNamesOnlyAUnitEveryCellCarries(t *testing.T) {
	for _, tc := range units.LabelConformance {
		if tc.Unit == "" {
			continue
		}
		for _, v := range []float64{0.45, 37.5, 313, 3661, 90000, 1234.5, 1.82e6} {
			cell := units.Format(v, tc.Unit)
			// An SI prefix goes *before* the symbol — "1.82 Mm" still ends in
			// "m" — which is exactly why a scaled column can keep its base
			// symbol in the header and a duration column cannot.
			named := strings.HasSuffix(cell, tc.Label)
			if tc.Label == "Time" {
				if named {
					t.Errorf("Label(%q) = %q but Format(%v, %q) = %q ends in it: the header should name the unit", tc.Unit, tc.Label, v, tc.Unit, cell)
				}
				continue
			}
			if !named {
				t.Errorf("Label(%q) = %q but Format(%v, %q) = %q does not end in it", tc.Unit, tc.Label, v, tc.Unit, cell)
			}
		}
	}
}

// The canonical separators are en-US's, and they are the *fallback* — the form
// a reader with no JavaScript keeps. The site re-renders them through
// Intl.NumberFormat, so what this asserts is that the fallback is a pair of
// characters people actually write, and not the U+202F narrow no-break space
// this used to bake in for everybody.
func TestCanonicalSeparators(t *testing.T) {
	got := units.Format(1234567, "")
	if got != "1,234,567" {
		t.Errorf("Format(1234567) = %q, want %q", got, "1,234,567")
	}
	if strings.ContainsAny(got, " \u00a0\u202f") {
		t.Errorf("Format(1234567) = %q, want no space of any width between the groups", got)
	}
	if got := units.Format(48750, units.Pascals); got != "48.8 kPa" {
		t.Errorf("Format(48750, Pa) = %q, want %q", got, "48.8 kPa")
	}
}

// Split is what the server-rendered site hands Intl.NumberFormat, so every
// invariant a browser relies on is asserted here rather than in JavaScript:
// Head+Tail is Format exactly, Number is the *scaled* value the reader sees,
// and Decimals is the post-trim count — a pre-trim count would put "1.820" back
// on the page.
func TestSplitDescribesWhatTheBrowserReRenders(t *testing.T) {
	for _, tc := range []struct {
		value    float64
		unit     string
		head     string
		tail     string
		number   float64
		decimals int
		isNumber bool
	}{
		{1820000, units.Metres, "1.82", " Mm", 1.82, 2, true},
		{7799, units.MetresSec, "7,799", " m/s", 7799, 0, true},
		{1234567, "", "1,234,567", "", 1234567, 0, true},
		{48000000, units.Joules, "48", " MJ", 48, 0, true},
		{6, "RUDs", "6", " RUDs", 6, 0, true},
		{-214.4, "", "-214", "", -214, 0, true},
		{0.002, "", "0.002", "", 0.002, 3, true},
		// Rule 5 under a minute is still one number, so it localises.
		{37500, units.Millis, "37.5", " s", 37.5, 1, true},
		{450, units.Millis, "450", " ms", 450, 0, true},
		{0, units.Seconds, "0", " s", 0, 0, true},
		// A two-component duration is not: it is two numbers and a layout, and
		// no component can reach a thousand.
		{2.1e7, units.Seconds, "243d 01h", "", 0, 0, false},
		{-3661, units.Seconds, "-1h 01m", "", 0, 0, false},
		// Neither is the not-a-number.
		{math.NaN(), units.Metres, "—", "", 0, 0, false},
	} {
		p := units.Split(tc.value, tc.unit)
		if p.Head != tc.head || p.Tail != tc.tail {
			t.Errorf("Split(%v, %q) = %q + %q, want %q + %q", tc.value, tc.unit, p.Head, p.Tail, tc.head, tc.tail)
		}
		if got := p.Head + p.Tail; got != units.Format(tc.value, tc.unit) {
			t.Errorf("Split(%v, %q) reassembles to %q, but Format says %q", tc.value, tc.unit, got, units.Format(tc.value, tc.unit))
		}
		if p.IsNumber != tc.isNumber {
			t.Errorf("Split(%v, %q).IsNumber = %v, want %v", tc.value, tc.unit, p.IsNumber, tc.isNumber)
		}
		if !tc.isNumber {
			continue
		}
		if p.Number != tc.number {
			t.Errorf("Split(%v, %q).Number = %v, want %v", tc.value, tc.unit, p.Number, tc.number)
		}
		if p.Decimals != tc.decimals {
			t.Errorf("Split(%v, %q).Decimals = %d, want %d", tc.value, tc.unit, p.Decimals, tc.decimals)
		}
	}
}

// The number a browser re-renders from must round-trip to the text a reader
// without one sees. If it did not, the two would disagree in the last digit
// depending on whether a script ran.
func TestSplitNumberRoundTripsToItsOwnText(t *testing.T) {
	for _, tc := range units.Conformance {
		p := units.Split(tc.Value, tc.Unit)
		if !p.IsNumber {
			continue
		}
		want := strings.ReplaceAll(p.Head, units.GroupSeparator, "")
		got := strconv.FormatFloat(p.Number, 'f', p.Decimals, 64)
		if got != want {
			t.Errorf("Split(%v, %q): Number %v at %d decimals is %q, but Head is %q",
				tc.Value, tc.Unit, p.Number, p.Decimals, got, want)
		}
	}
}

func TestForKeyReadsTheSuffix(t *testing.T) {
	for key, want := range map[string]string{
		// The trap: `_ms` is metres per second in a payload, but the board unit
		// string "ms" is milliseconds. Only this function knows the difference.
		"speed_ms":         units.MetresSec,
		"surface_speed_ms": units.MetresSec,
		"fastest_ms":       units.MetresSec,
		"accel_ms2":        units.MetresSec2,
		"altitude_m":       units.Metres,
		"ap_m":             units.Metres,
		"travelled_m":      units.Metres,
		"energy_j":         units.Joules,
		"peak_q_pa":        units.Pascals,
		"max_q_pa":         units.Pascals,
		"dyn_pressure_pa":  units.Pascals,
		"duration_s":       units.Seconds,
		"mission_time_s":   units.Seconds,
		"t1_sim":           units.Seconds,
		"sim_t":            units.Seconds,
		"mass_kg":          units.Kilograms,
		"inc_deg":          units.Degrees,
		"peak_g":           units.Gs,
		// Unrecognised keys get no unit rather than a wrong one.
		"body":   "",
		"career": "",
		"ecc":    "",
		"n":      "",
	} {
		if got := units.ForKey(key); got != want {
			t.Errorf("ForKey(%q) = %q, want %q", key, got, want)
		}
	}
}

// Every unit any board publishes must render without falling through to
// something that looks like a bug — a board added later with a new label gets
// "12 whatevers", never "12 " or a panic.
func TestEveryBoardUnitRenders(t *testing.T) {
	for _, unit := range []string{"m/s", "g", "tumbles", "RUDs", "orbits", "bodies", "dockings", "stagings", "kittens", "m", "ms"} {
		got := units.Format(1234.5, unit)
		if got == "" || strings.HasSuffix(got, " ") {
			t.Errorf("Format(1234.5, %q) = %q", unit, got)
		}
	}
}
